toolchain/libpfm.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2018 IBM.
+# Author: Harish <harish@linux.vnet.ibm.com>
+
+
+import os
+from avocado.utils import process
+from avocado import Test
+from avocado import main
+from avocado.utils import build, distro, archive
+from avocado.utils.software_manager import SoftwareManager
+
+
+class Libpfm(Test):
+
+    '''
+    This testcase make use of testsuite provided by the
+    source package, source files are downloaded and compiled
+    '''
+
+    def setUp(self):
+        softm = SoftwareManager()
+
+        pkgs = ['gcc', 'make']
+        if distro.detect().name in ['SuSE', 'Ubuntu']:
+            pkgs.extend(['libpfm4'])
+        else:
+            pkgs.extend(['libpfm'])
+
+        for pkg in pkgs:
+            if not softm.check_installed(pkg) and not softm.install(pkg):
+                self.cancel("%s is needed for the test to be run" % pkg)
+        test_type = self.params.get('type', default='upstream')
+
+        if test_type == 'upstream':
+            tarball = self.fetch_asset(
```

Comment:
In the case of upstream, it would be good to fetch the current code from the `git tree`
```
https://git.code.sf.net/p/perfmon2/libpfm4
```

Reply:
@nasastry I had that before. Looks to be a lag in framework where with larger-git-repo (eg. linux ) & slower-server-response (in this case),  `git fetch` times out. So, used the download link from the maintaining website.

Reply:
@harish-24 sorry for late response but it's not taking much time (hardly 3 sec)

```
# time git clone https://git.code.sf.net/p/perfmon2/libpfm4
Cloning into 'libpfm4'...
remote: Counting objects: 6827, done.
remote: Compressing objects: 100% (1924/1924), done.
remote: Total 6827 (delta 5201), reused 6419 (delta 4888)
Receiving objects: 100% (6827/6827), 2.28 MiB | 1.44 MiB/s, done.
Resolving deltas: 100% (5201/5201), done.

real	0m3.522s
user	0m1.145s
sys	0m0.082s
```

Reply:
@nasastry Can you try the same with avocado and let know? Because at the time I was testing, fetch timed-out.

Reply:
@harish-24 with avocado it not worked i already tried

Reply:
@harish-24 No issues from my side.